// nativescript-speech-recognition/speech-recognition.android.js
"use strict";
var application = require("application");
var utils = require("utils/utils");
var SpeechRecognition = (function () {
    function SpeechRecognition() {
        var self = this;
        application.android.on(application.AndroidApplication.activityRequestPermissionsEvent, function (args) {
            for (var i = 0; i < args.permissions.length; i++) {
                if (args.grantResults[i] === android.content.pm.PackageManager.PERMISSION_DENIED) {
                    if (self.onPermissionRejected) {
                        self.onPermissionRejected("Please allow access to the Microphone and try again.");
                    }
                    else {
                        console.log("Please allow access to the Microphone and try again. (tip: pass in a reject to receive this message in your app)");
                    }
                    return;
                }
            }
            if (self.onPermissionGranted) {
                self.onPermissionGranted();
            }
        });
    }
    SpeechRecognition.prototype.available = function () {
        return new Promise(function (resolve, reject) {
            resolve(android.speech.SpeechRecognizer.isRecognitionAvailable(application.android.context));
        });
    };
    ;
    SpeechRecognition.prototype.startListening = function (options) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            var onPermissionGranted = function () {
                var loopHandler = new android.os.Handler(android.os.Looper.getMainLooper());
                loopHandler.post(new java.lang.Runnable({
                    run: function () {
                        _this.recognizer = android.speech.SpeechRecognizer.createSpeechRecognizer(application.android.context);
                        _this.recognizer.setRecognitionListener(new android.speech.RecognitionListener({
                            onReadyForSpeech: function (params) {
                                resolve(true);
                            },
                            onBeginningOfSpeech: function () {
                            },
                            onRmsChanged: function (rmsdB) {
                            },
                            onBufferReceived: function (buffer) {
                            },
                            onEndOfSpeech: function () {
                            },
                            onError: function (error) {
                            },
                            onResults: function (results) {
                                this.sendBackResults(results);
                            },
                            onPartialResults: function (partialResults) {
                                if (options.returnPartialResults) {
                                    this.sendBackResults(partialResults);
                                }
                            },
                            sendBackResults: function (results) {
                                var transcripts = results.getStringArrayList(android.speech.SpeechRecognizer.RESULTS_RECOGNITION);
                                if (!transcripts.isEmpty()) {
                                    for (var i = 0; i < transcripts.size(); i++) {
                                        var transcript = transcripts.get(i);
                                        console.log("--- trans " + i + ": " + transcript);
                                    }
                                    options.onResult({
                                        text: transcripts[0],
                                        finished: true
                                    });
                                }
                            },
                            onEvent: function (eventType, params) {
                            }
                        }));
                    }
                }));
                var intent = new android.content.Intent(android.speech.RecognizerIntent.ACTION_RECOGNIZE_SPEECH);
                intent.putExtra(android.speech.RecognizerIntent.EXTRA_LANGUAGE_MODEL, android.speech.RecognizerIntent.LANGUAGE_MODEL_FREE_FORM);
                intent.putExtra(android.speech.RecognizerIntent.EXTRA_CALLING_PACKAGE, "voice.recognition.test");
                intent.putExtra(android.speech.RecognizerIntent.EXTRA_LANGUAGE, "en");
                intent.putExtra(android.speech.RecognizerIntent.EXTRA_MAX_RESULTS, 100);
                loopHandler.post(new java.lang.Runnable({
                    run: function () {
                        _this.recognizer.startListening(intent);
                    }
                }));
            };
            if (!_this.wasPermissionGranted()) {
                _this.requestPermission(onPermissionGranted, reject);
                return;
            }
            onPermissionGranted();
        });
    };
    SpeechRecognition.prototype.stopListening = function () {
        var _this = this;
        return new Promise(function (resolve, reject) {
            var loopHandler = new android.os.Handler(android.os.Looper.getMainLooper());
            loopHandler.post(new java.lang.Runnable({
                run: function () {
                    _this.recognizer.stopListening();
                    resolve();
                }
            }));
        });
    };
    SpeechRecognition.prototype.wasPermissionGranted = function () {
        var hasPermission = android.os.Build.VERSION.SDK_INT < 23;
        if (!hasPermission) {
            hasPermission = android.content.pm.PackageManager.PERMISSION_GRANTED ===
                android.support.v4.content.ContextCompat.checkSelfPermission(utils.ad.getApplicationContext(), android.Manifest.permission.RECORD_AUDIO);
        }
        return hasPermission;
    };
    ;
    SpeechRecognition.prototype.requestPermission = function (onPermissionGranted, reject) {
        this.onPermissionGranted = onPermissionGranted;
        this.onPermissionRejected = reject;
        android.support.v4.app.ActivityCompat.requestPermissions(application.android.foregroundActivity, [android.Manifest.permission.RECORD_AUDIO], 444);
    };
    return SpeechRecognition;
}());
exports.SpeechRecognition = SpeechRecognition;
